src/components/ImagePreviewer.js

```diff
@@ -0,0 +1,176 @@
+import { stringify } from "querystring";
+import { decode } from "base58";
+
+const noop = () => {};
+
+export const of = src => Promise.resolve({ src });
+
+export const makeSrcResolver = ({ src }) =>
+  resolvers.reduceRight((acc, resolver) => {
+    if (!acc && resolver.test(src)) {
+      return resolver.request(src);
+    }
+    return acc;
+  }, false);
+
+export const makeImageDOMResolver = ({ src }) =>
+  new Promise((resolve, reject) => {
+    const img = new Image();
+    img.onload = () =>
+      resolve({
+        src,
+        height: img.height
+      });
+    img.onerror = reject;
+    img.src = src;
+  });
+
+export class ImagePreviewer extends React.PureComponent {
+  state = {
+    pending: undefined,
+    value: undefined
+  };
+
+  componentDidMount() {
+    this.handleStart();
+  }
+
+  componentDidUpdate(prevProps) {
+    if (this.props.request !== prevProps.request) {
+      this.handleStart();
+    }
+  }
+
+  handleStart(props) {
+    this.setState((state, { request }) => {
+      request.then(this.handleResolve, noop);
+      return {
+        pending: request,
+        value: undefined
+      };
+    });
+  }
+
+  handleResolve = value => {
+    this.setState(({ pending }, { request }) => {
+      if (pending !== request) {
+        return;
+      }
+      return { value };
+    });
+  };
+
+  render() {
+    return this.props.render({
```

Comment:
Can it be replaced with dynamic type?
https://reactjs.org/docs/jsx-in-depth.html#choosing-the-type-at-runtime

Reply:
Yes. I just choose render props over dynamic type due to the reasons:

- We're unsure whether the extracted types will be reused again, so just keep it as simple as possible.
- Render function don't create an extra layer of react components in the render tree

Reply:
Ack.